### ISSUE TITLE
Default keybindings conflict with Emacs

### ISSUE THREAD

**Issue by andyleejordan** [ArcticaProject/nx-libs#800]
A lot of the default keybindings in [keystrokes.cfg](https://github.com/ArcticaProject/nx-libs/blob/3.6.x/etc/keystrokes.cfg) conflict with the default bindings of the _world's most popular text editor_, Emacs. Specifically, I can't kill s-expressions (Ctrl-Alt-K)!

I tracked it down to this default bindings for [reread_keystrokes](https://github.com/ArcticaProject/nx-libs/blob/4ce890758035078bb1ca133cc4584ca02db8ad6d/etc/keystrokes.cfg#L27), which is not mentioned on the [wiki page](https://wiki.x2go.org/doku.php/wiki:advanced:nx-keyboard-shortcuts).

I am using X2Go as a client on Windows 10, with a Linux server. Almost everything works flawlessly and amazingly fast, but I can't find documentation on how to override these bindings for the _Windows_ client, as `~/.x2go/config/keystrokes.cfg` doesn't exist. I tried creating it, but it's not being respected; I also tried pointing to the one I created with `NXAGENT_KEYSTROKEFILE` but still no luck.

How can I change these bindings for the Windows client?

Thanks!

**Comment by uli42** [comment]
The keystrokes config is a _ server side_ config file. You need to adapt
/etc/x2go/keystrokes.cfg or place one into $HOME/.x2go/keystrokes.cfg.

Uli

Andrew Schwartzmeyer <notifications@github.com> schrieb am Di., 30. Apr.
2019, 18:58:

**Comment by uli42** [comment]
On Tue, Apr 30, 2019 at 7:14 PM Ulrich Sibiller
<ulrich.sibiller@gmail.com> wrote:
>
> The keystrokes config is a _ server side_ config file. You need to adapt /etc/x2go/keystrokes.cfg or place one into $HOME/.x2go/keystrokes.cfg


BTW: /etc/x2go/keystrokes.cfg is a bit outdated (and buggy), better
copy the one from /etc/nxagent/ to /etc/x2go

Uli

**Comment by andyleejordan** [comment]
@uli42 Thank you! Unfortunately, I am still not quite able to get `Ctrl-Alt-k` to work. `Ctrl-Alt-Shift-k` works (as in, Emacs receives it over x2go).

I've edited both `/etc/x2go/keystrokes.cfg` and `~/.x2go/keystrokes.cfg` and restarted the server (well, the whole machine), but can't get `Ctrl-Alt-k` not to be intercepted.

But I'm no longer convinced x2go is ever getting the keystroke. I noticed when testing `Ctrl-Alt-t` to close the session that despite it working to close the session, the keystroke is actually also sent inside the session (as the corresponding Emacs command is called right before the session is closed)...hm time for more debugging.

**Comment by andyleejordan** [comment]
That said...I actually can't get any change to either of those files to take place...maybe I've done something wrong.

The server is Ubuntu 18.04, running:

```
> x2goversion 
x2goserver: 4.1.0.0
x2goserver-common: 4.1.0.0
x2goserver-extensions: 4.1.0.0
x2goserver-fmbindings: 4.1.0.0
x2goserver-printing: 4.1.0.0
x2goserver-x2goagent: 3.5.99.16
x2goserver-xsession: 4.1.0.0
```

**Comment by uli42** [comment]
On Tue, Apr 30, 2019 at 9:07 PM Andrew Schwartzmeyer
<notifications@github.com> wrote:
>
> @uli42 Thank you! Unfortunately, I am still not quite able to get Ctrl-Alt-k to work. Ctrl-Alt-Shift-k works (as in, Emacs receives it over x2go).
>
> I've edited both /etc/x2go/keystrokes.cfg and ~/.x2go/keystrokes.cfg and restarted the server (well, the whole machine), but can't get Ctrl-Alt-K not to be intercepted.

Sorry, my fault. The personal file is ~/.x2go/config/keystrokes.cfg.
If it exists, the one in /etc will be ignored.

You can see the active keystrokes (and the used config file) in
~/.x2go/C-<session-name>/session.log

> But I'm no longer convinced x2go is ever getting the keystroke. I noticed when testing Ctrl-Alt-t to close the session that despite it working to close the session, the keystroke is actually also sent inside the session (as the corresponding Emacs command is called right before the session is closed)...hm time for more debugging

Interesting. I do not see that behaviour on a Linux client. For
testing can please disable the keystroke by configuring it like this:

<keystroke action="ignore" Control="1" AltMeta="1" key="k" />

and see what happens.

Uli

**Comment by uli42** [comment]
On Tue, Apr 30, 2019 at 9:16 PM Andrew Schwartzmeyer
<notifications@github.com> wrote:
>
> That said...I actually can't get any change to either of those files to take place...maybe I've done something wrong.
>
> The server is Ubuntu 18.04, running:
>
> > x2goversion
> x2goserver: 4.1.0.0
> x2goserver-common: 4.1.0.0
> x2goserver-extensions: 4.1.0.0
> x2goserver-fmbindings: 4.1.0.0
> x2goserver-printing: 4.1.0.0
> x2goserver-x2goagent: 3.5.99.16
> x2goserver-xsession: 4.1.0.0

There are newer versions, please upgrade at least nxagent to
3.5.99.20. I want to avoid hunting bugs that have already been fixed.

BTW: the most current keystroke documentation is in
/usr/share/doc/nxagent/README.keystrokes.gz (provided you have the
current nxagent package).

Uli

**Comment by andyleejordan** [comment]
Draft: I installed the PPA and updated the package and noticed `/etc/nxagent/keystrokes.cfg` getting updated (which I glossed over in your message, didn't realize that would actually exist on my system).

Ah, see your messages, thank you! I am reading `Keystroke.c` and figuring that out 😄 thanks for `action=ignore` will try!

**Comment by andyleejordan** [comment]
Updated packages:

```
> nxagent -version
NXAGENT - Version 3.5.99.20
> x2goversion
: 4.1.0.3
: 4.1.0.3
: 4.1.0.3
: 4.1.0.3
: 4.1.0.3
: 3.5.99.20
: 4.1.0.3
```

**Comment by andyleejordan** [comment]
Here's my log:

```

running as X2Go Agent

running as X2Go Agent

running as X2Go Agent

running as X2Go Agent

running as X2Go Agent

running as X2Go Agent

running as X2Go Agent

running as X2Go Agent

NXAGENT - Version 3.5.99.20

Copyright (c) 2001, 2011 NoMachine (http://www.nomachine.com)
Copyright (c) 2008-2014 Oleksandr Shneyder <o.shneyder@phoca-gmbh.de>
Copyright (c) 2011-2016 Mike Gabriel <mike.gabriel@das-netzwerkteam.de>
Copyright (c) 2014-2016 Ulrich Sibiller <uli42@gmx.de>
Copyright (c) 2014-2016 Mihai Moldovan <ionic@ionic.de>
Copyright (c) 2015-2016 Qindel Group (http://www.qindel.com)
See https://github.com/ArcticaProject/nx-libs for more information.

Info: Agent running with pid '2613'.
Session: Starting session at 'Tue Apr 30 19:43:02 2019'.
Info: Proxy running in client mode with pid '2613'.
Info: Using errors file '/tmp/.x2go-andschwa/C-andschwa-50-1556653381_stDXFCE_dp32/session.log'.
Info: Using stats file '/tmp/.x2go-andschwa/C-andschwa-50-1556653381_stDXFCE_dp32/C-andschwa-50-1556653381_stDXFCE_dp32/stats'.
Loop: WARNING! Unrecognized session type 'unix-kde-depth_32'. Assuming agent session.
Warning: Unrecognized session type 'unix-kde-depth_32'. Assuming agent session.
Info: Waiting for connection from 'localhost' on socket 'tcp:*:47138'.
Info: Accepted connection from '127.0.0.1'.
Info: Connection with remote proxy completed.
Info: Using LAN link parameters 1536/24/1/0.
Info: Using agent parameters 5000/0/50/0/0.
Info: Using pack method '16m-jpeg-9' with session 'unix-kde-depth_32'.
Info: Not using NX delta compression.
Info: Not using ZLIB data compression.
Info: Not using ZLIB stream compression.
Info: Not using a persistent cache.
Info: Listening to X11 connections on display ':50'.
Info: Established X client connection.
Info: Using shared memory parameters 1/1/0/0K.
Info: Using alpha channel in render extension.
Info: Not using local device configuration changes.
Info: using keystrokes file '/home/andschwa/.x2go/config/keystrokes.cfg'
Current known keystrokes:
  ignore Ctrl+Alt+k
Info: keyboard blocking directory created: '/home/andschwa/.x2go/C-andschwa-50-1556653381_stDXFCE_dp32/keyboard'
Session: Session started at 'Tue Apr 30 19:43:08 2019'.
Info: Screen [0] resized to geometry [2560x1440] fullscreen [0].
```

**Comment by andyleejordan** [comment]
Ah ha! With `~/.x2go/config/keystrokes.cfg` created, with just:

```
<!DOCTYPE NXKeystroke>
<keystrokes>
</keystrokes>
```

Then Emacs in the session receives `C-M-k` just as I needed. Funny thing was with `action="ignore"` it still didn't receive it.

Thank you so much for your help! It wasn't clear to me that the config file was _server side_, and if I did remember to try `~/.x2go/config/keystrokes.cfg` with an almost empty config (as I certainly tried the others), I must not have restarted the service.

Everything is working now. Thank you.

**Comment by andyleejordan** [comment]
Oh wait what. I rebooted again to double-check. And with the log now I see why I had so much trouble earlier!

That's not right. It's trying to read the wrong file!

```
Warning: Cannot read keystroke file '/home/andschwa/.nx/config/keystrokes.cfg'.
Info: using keystrokes file '/etc/nxagent/keystrokes.cfg'
Current known keystrokes:
  close_session Ctrl+Alt+t
  switch_all_screens Ctrl+Alt+f
  fullscreen Ctrl+Shift+Alt+f
  minimize Ctrl+Alt+m
  defer Ctrl+Alt+e
  ignore Ctrl+Alt+BackSpace
  force_synchronization Ctrl+Alt+j
  resize Ctrl+Alt+r
  viewport_move_left Ctrl+Shift+Alt+Left
  viewport_move_left Ctrl+Shift+Alt+KP_Left
  viewport_move_up Ctrl+Shift+Alt+Up
  viewport_move_up Ctrl+Shift+Alt+KP_Up
  viewport_move_right Ctrl+Shift+Alt+Right
  viewport_move_right Ctrl+Shift+Alt+KP_Right
  viewport_move_down Ctrl+Shift+Alt+Down
  viewport_move_down Ctrl+Shift+Alt+KP_Down
  viewport_scroll_left Ctrl+Alt+Left
  viewport_scroll_left Ctrl+Alt+KP_Left
  viewport_scroll_up Ctrl+Alt+Up
  viewport_scroll_up Ctrl+Alt+KP_Up
  viewport_scroll_right Ctrl+Alt+Right
  viewport_scroll_right Ctrl+Alt+KP_Right
  viewport_scroll_down Ctrl+Alt+Down
  viewport_scroll_down Ctrl+Alt+KP_Down
  reread_keystrokes Ctrl+Alt+k
```

**Comment by andyleejordan** [comment]
Trying to repro right now just to triple-check, but it looks like it connects and runs as x2go agent, then it hits an issue and restarts itself, and the restarted session no longer runs as x2go agent and so swaps its logic to the nxagent logic, and so reads the _other_ file. For now I'm just going to set both up with overrides to remove all bindings, but sure is odd that the automatic restart it did doesn't maintain the semantics.

**Comment by andyleejordan** [comment]
Well it didn't repro, and I dumbly didn't save the part of the log where it said FATAL and then looked like it rebooted itself. If it happens again I'll let you know. With x2go and nxagent confs set with my overrides, I am good to go, so thank you.

**Comment by uli42** [comment]
Puuh, this looks like a bug. But I never ever saw nxagent restarting itself. Maybe that's a feature of x2goserver I am not aware of: use nxagent if x2goagent fails for some reason.

However: Just to be sure, please answer these questions (again):
- Does it work if the config file exists and does not contain any config line?
- Does it work if ctrl-alt-k is mapped to ignore?
- Does it work without any config file?

(re-opening until these questions are clear)

**Comment by uli42** [comment]
On Tue, Apr 30, 2019 at 10:05 PM Andrew Schwartzmeyer
<notifications@github.com> wrote:
>
> Trying to repro right now just to triple-check, but it looks like it connects and runs as x2go agent, then it hits an issue and restarts itself, and the restarted session no longer runs as x2go agent and so swaps its logic to the nxagent logic, and so reads the other file. For now I'm just going to set both up with overrides to remove all bindings, but sure is odd that the automatic restarst it did doesn't maintain the semantics.

Puuh, this looks like a bug. But I never ever saw nxagent restarting
itself. Maybe that's a feature of x2goserver I am not aware of: use
nxagent if x2goagent fails for some reason.

**Comment by andyleejordan** [comment]
Sorry to be getting back to you so late. To clarify, "it works" would be applications used in X2GO receive the key chord.

> Does it work if the config file exists and does not contain any config line?

No.

```
/home/andschwa/.x2go/config/keystrokes.cfg:1: parser error : Document is empty
 
^
Warning: could not read/parse keystrokes file '/home/andschwa/.x2go/config/keystrokes.cfg'
Info: Using builtin keystrokes.
```

(Side note, I am also having issues getting copy/paste between Windows and Linux working, so getting these logs is annoying 😓.)

> Does it work if ctrl-alt-k is mapped to ignore?

No. It correctly is mapped to "ignore" but it's not then passed along to any applications.

```
Info: using keystrokes file '/home/andschwa/.x2go/config/keystrokes.cfg'
Current known keystrokes:
  ignore Ctrl+Alt+k
```

> Does it work without any config file?

No. Without any (user) config file it just defaults to the known keystrokes, overriding Ctrl+Alt+k with the reread configuration or whatever.

What does work:

```
<!DOCTYPE NXKeystroke>
<keystrokes>
</keystrokes>
```

Also, I've removed by NX config file, so if I see that earlier issue crop up again I will grab the log for you.

**Comment by uli42** [comment]
Thx. I will investigate that ignore issue (which should better be called
pass-through).

Uli

Andrew Schwartzmeyer <notifications@github.com> schrieb am Di., 7. Mai
2019, 01:47:

**Comment by andyleejordan** [comment]
Thanks @uli42, I appreciate it!

**Comment by andyleejordan** [comment]
@uli42 it happened again!

```

running as X2Go Agent

running as X2Go Agent

running as X2Go Agent

running as X2Go Agent

running as X2Go Agent

running as X2Go Agent

running as X2Go Agent

running as X2Go Agent

NXAGENT - Version 3.5.99.20

Copyright (c) 2001, 2011 NoMachine (http://www.nomachine.com)
Copyright (c) 2008-2014 Oleksandr Shneyder <o.shneyder@phoca-gmbh.de>
Copyright (c) 2011-2016 Mike Gabriel <mike.gabriel@das-netzwerkteam.de>
Copyright (c) 2014-2016 Ulrich Sibiller <uli42@gmx.de>
Copyright (c) 2014-2016 Mihai Moldovan <ionic@ionic.de>
Copyright (c) 2015-2016 Qindel Group (http://www.qindel.com)
See https://github.com/ArcticaProject/nx-libs for more information.

Info: Agent running with pid '16549'.
Session: Starting session at 'Mon May  6 23:20:40 2019'.
Info: Proxy running in client mode with pid '16549'.
Info: Using errors file '/tmp/.x2go-andschwa/C-andschwa-50-1557184839_stDXFCE_dp32/session.log'.
Info: Using stats file '/tmp/.x2go-andschwa/C-andschwa-50-1557184839_stDXFCE_dp32/C-andschwa-50-1557184839_stDXFCE_dp32/stats'.
Loop: WARNING! Unrecognized session type 'unix-kde-depth_32'. Assuming agent session.
Warning: Unrecognized session type 'unix-kde-depth_32'. Assuming agent session.
Info: Waiting for connection from 'localhost' on socket 'tcp:*:32629'.
Info: Accepted connection from '127.0.0.1'.
Info: Connection with remote proxy completed.
Info: Using LAN link parameters 1536/24/1/0.
Info: Using agent parameters 5000/0/50/0/0.
Info: Using pack method '16m-jpeg-9' with session 'unix-kde-depth_32'.
Info: Not using NX delta compression.
Info: Not using ZLIB data compression.
Info: Not using ZLIB stream compression.
Info: Not using a persistent cache.
Info: Listening to X11 connections on display ':50'.
Info: Established X client connection.
Info: Using shared memory parameters 1/1/0/0K.
Info: Using alpha channel in render extension.
Info: Not using local device configuration changes.
Info: using keystrokes file '/home/andschwa/.x2go/config/keystrokes.cfg'
Current known keystrokes:
  ignore Ctrl+Alt+k
Info: keyboard blocking directory created: '/home/andschwa/.x2go/C-andschwa-50-1557184839_stDXFCE_dp32/keyboard'
Session: Session started at 'Mon May  6 23:20:45 2019'.
Info: Screen [0] resized to geometry [2880x1620] fullscreen [0].
Info: Screen [0] resized to geometry [2880x1620] fullscreen [0].
Info: Screen [0] resized to geometry [2880x2056] fullscreen [0].
Info: Screen [0] resized to geometry [2880x2056] fullscreen [0].
Info: Screen [0] resized to geometry [2880x1893] fullscreen [0].
Info: Screen [0] resized to geometry [2880x1893] fullscreen [0].
Info: Screen [0] resized to geometry [2880x2020] fullscreen [0].
Info: Screen [0] resized to geometry [2880x2020] fullscreen [0].
Info: Screen [0] resized to geometry [2880x1893] fullscreen [0].
Info: Screen [0] resized to geometry [2880x2020] fullscreen [0].
Info: Screen [0] resized to geometry [2880x2020] fullscreen [0].
Info: Screen [0] resized to geometry [2880x2100] fullscreen [0].
Info: Screen [0] resized to geometry [2880x2100] fullscreen [0].
Info: Screen [0] resized to geometry [2880x2020] fullscreen [0].
Info: Screen [0] resized to geometry [3840x2035] fullscreen [0].
Session: Suspending session at 'Tue May  7 01:37:57 2019'.
Info: Waiting the cleanup timeout to complete.
Session: Session suspended at 'Tue May  7 01:37:58 2019'.
Xlib: resource ID allocation space exhausted!
Session: Resuming session at 'Tue May  7 21:58:41 2019'.
Info: Proxy running in client mode with pid '16549'.
Info: Using errors file '/tmp/.x2go-andschwa/C-andschwa-50-1557184839_stDXFCE_dp32/session.log'.
Info: Using stats file '/tmp/.x2go-andschwa/C-andschwa-50-1557184839_stDXFCE_dp32/C-andschwa-50-1557184839_stDXFCE_dp32/stats'.
Loop: WARNING! Unrecognized session type 'unix-kde-depth_32'. Assuming agent session.
Warning: Unrecognized session type 'unix-kde-depth_32'. Assuming agent session.
Info: Waiting for connection from 'localhost' on socket 'tcp:*:32629'.
Info: Accepted connection from '127.0.0.1'.
Info: Connection with remote proxy completed.
Info: Using LAN link parameters 1536/24/1/0.
Info: Using agent parameters 5000/0/50/0/0.
Info: Using pack method '16m-jpeg-9' with session 'unix-kde-depth_32'.
Info: Not using NX delta compression.
Info: Not using ZLIB data compression.
Info: Not using ZLIB stream compression.
Info: Not using a persistent cache.
Info: Listening to X11 connections on display ':50'.
Info: Established X client connection.
Info: Using shared memory parameters 1/1/0/0K.
Info: Not using local device configuration changes.
Info: re-reading keystrokes configuration
Warning: Cannot read keystroke file '/home/andschwa/.nx/config/keystrokes.cfg'.
Info: using keystrokes file '/etc/nxagent/keystrokes.cfg'
Current known keystrokes:
  close_session Ctrl+Alt+t
  switch_all_screens Ctrl+Alt+f
  fullscreen Ctrl+Shift+Alt+f
  minimize Ctrl+Alt+m
  defer Ctrl+Alt+e
  ignore Ctrl+Alt+BackSpace
  force_synchronization Ctrl+Alt+j
  resize Ctrl+Alt+r
  viewport_move_left Ctrl+Shift+Alt+Left
  viewport_move_left Ctrl+Shift+Alt+KP_Left
  viewport_move_up Ctrl+Shift+Alt+Up
  viewport_move_up Ctrl+Shift+Alt+KP_Up
  viewport_move_right Ctrl+Shift+Alt+Right
  viewport_move_right Ctrl+Shift+Alt+KP_Right
  viewport_move_down Ctrl+Shift+Alt+Down
  viewport_move_down Ctrl+Shift+Alt+KP_Down
  viewport_scroll_left Ctrl+Alt+Left
  viewport_scroll_left Ctrl+Alt+KP_Left
  viewport_scroll_up Ctrl+Alt+Up
  viewport_scroll_up Ctrl+Alt+KP_Up
  viewport_scroll_right Ctrl+Alt+Right
  viewport_scroll_right Ctrl+Alt+KP_Right
  viewport_scroll_down Ctrl+Alt+Down
  viewport_scroll_down Ctrl+Alt+KP_Down
  reread_keystrokes Ctrl+Alt+k
Session: Session resumed at 'Tue May  7 21:59:10 2019'.
Info: Screen [0] resized to geometry [2880x1620] fullscreen [0].
Info: Screen [0] resized to geometry [2880x1620] fullscreen [0].
Info: Screen [0] resized to geometry [3032x2020] fullscreen [0].
Info: Screen [0] resized to geometry [3032x2020] fullscreen [0].
Info: Screen [0] resized to geometry [2880x1620] fullscreen [0].
Session: Suspending session at 'Tue May  7 22:30:41 2019'.
Info: Waiting the cleanup timeout to complete.
Loop: PANIC! No shutdown of proxy link performed by remote proxy.
Error: Connection with remote peer broken.
Error: Please check the state of your network and retry.
Session: Session suspended at 'Tue May  7 22:30:41 2019'.
Session: Resuming session at 'Tue May  7 22:32:08 2019'.
Info: Proxy running in client mode with pid '16549'.
Info: Using errors file '/tmp/.x2go-andschwa/C-andschwa-50-1557184839_stDXFCE_dp32/session.log'.
Info: Using stats file '/tmp/.x2go-andschwa/C-andschwa-50-1557184839_stDXFCE_dp32/C-andschwa-50-1557184839_stDXFCE_dp32/stats'.
Loop: WARNING! Unrecognized session type 'unix-kde-depth_32'. Assuming agent session.
Warning: Unrecognized session type 'unix-kde-depth_32'. Assuming agent session.
Info: Waiting for connection from 'localhost' on socket 'tcp:*:32629'.
Info: Accepted connection from '127.0.0.1'.
Info: Connection with remote proxy completed.
Info: Using LAN link parameters 1536/24/1/0.
Info: Using agent parameters 5000/0/50/0/0.
Info: Using pack method '16m-jpeg-9' with session 'unix-kde-depth_32'.
Info: Not using NX delta compression.
Info: Not using ZLIB data compression.
Info: Not using ZLIB stream compression.
Info: Not using a persistent cache.
Info: Listening to X11 connections on display ':50'.
Info: Established X client connection.
Info: Using shared memory parameters 1/1/0/0K.
Info: Not using local device configuration changes.
Info: re-reading keystrokes configuration
Warning: Cannot read keystroke file '/home/andschwa/.nx/config/keystrokes.cfg'.
Info: using keystrokes file '/etc/nxagent/keystrokes.cfg'
Current known keystrokes:
  close_session Ctrl+Alt+t
  switch_all_screens Ctrl+Alt+f
  fullscreen Ctrl+Shift+Alt+f
  minimize Ctrl+Alt+m
  defer Ctrl+Alt+e
  ignore Ctrl+Alt+BackSpace
  force_synchronization Ctrl+Alt+j
  resize Ctrl+Alt+r
  viewport_move_left Ctrl+Shift+Alt+Left
  viewport_move_left Ctrl+Shift+Alt+KP_Left
  viewport_move_up Ctrl+Shift+Alt+Up
  viewport_move_up Ctrl+Shift+Alt+KP_Up
  viewport_move_right Ctrl+Shift+Alt+Right
  viewport_move_right Ctrl+Shift+Alt+KP_Right
  viewport_move_down Ctrl+Shift+Alt+Down
  viewport_move_down Ctrl+Shift+Alt+KP_Down
  viewport_scroll_left Ctrl+Alt+Left
  viewport_scroll_left Ctrl+Alt+KP_Left
  viewport_scroll_up Ctrl+Alt+Up
  viewport_scroll_up Ctrl+Alt+KP_Up
  viewport_scroll_right Ctrl+Alt+Right
  viewport_scroll_right Ctrl+Alt+KP_Right
  viewport_scroll_down Ctrl+Alt+Down
  viewport_scroll_down Ctrl+Alt+KP_Down
  reread_keystrokes Ctrl+Alt+k
Session: Session resumed at 'Tue May  7 22:32:36 2019'.
Info: Screen [0] resized to geometry [2880x1620] fullscreen [0].
Info: Screen [0] resized to geometry [3232x2020] fullscreen [0].
Info: Screen [0] resized to geometry [3232x2020] fullscreen [0].
```

When it suspends and resumes the session, it comes back up as nxagent, not x2go, and so reads the nx config files, not x2go's.

ALSO, copy/paste for me is completely broken when it's the x2go agent, but if the session is suspended and resumed (and so it comes back as an nxagent), copy/paste works just fine.

I can repro this by starting a new session, checking that the log has it using the x2go configuration, and that copy/paste does not work, and then just closing the Windows client and re-opening it (leaving the server running), the log then shows it read the nxagent configuration, and copy/paste works (and I can keep doing this, hence the log showing nxagent resuming twice). If I instead, logout (killing the server) and reconnect, it comes back as x2go (of course).

**Comment by uli42** [comment]
I have identified the "problem": it works as designed ;-)

"ignore" is really meant for really ignoring the keystroke. So all you need
to do is remove the according line and you'll be fine.

I am not really satisfied with this. I am wondering if we need a way that
marks a keystroke as "passthrough" or if deleting the line is enough.

Uli

PS: I also found the initial reason for "ignore" was to block
ctrl-alt-backspace but nxagent does not even contain code zapping the
server when this keystroke is seen.

**Comment by andyleejordan** [comment]
> "ignore" is really meant for really ignoring the keystroke. So all you need to do is remove the according line and you'll be fine. I am not really satisfied with this. I am wondering if we need a way that marks a keystroke as "passthrough" or if deleting the line is enough.

In my case, deleting the line is enough (that is, have an almost-empty conf file), but we should probably document this.

But x2go agent resuming as nxagent, however, is definitely a (separate) bug.

**Comment by uli42** [comment]
On Wed, May 8, 2019 at 12:41 AM Andrew Schwartzmeyer <
notifications@github.com> wrote:

> Session: Suspending session at 'Tue May  7 01:37:57 2019'.
> Info: Waiting the cleanup timeout to complete.
> Session: Session suspended at 'Tue May  7 01:37:58 2019'.
> Xlib: resource ID allocation space exhausted!
> Session: Resuming session at 'Tue May  7 21:58:41 2019'.
>
>
I am wondering where that Xlib message is coming from.

I have tried replicating that behaviour but failed so far... Will try
further.

> When it suspends and resumes the session, it comes back up as nxagent, not
> x2go, and so reads the nx config files, not x2go's.
>
> ALSO, copy/paste for me is completely broken when it's the x2go agent, but
> if the session is suspended and resumed (and so it comes back as an
> nxagent), copy/paste works just fine.
>
> I can repro this by starting a new session, checking that the log has it
> using the x2go configuration, and that copy/paste does not work, and then
> just closing the Windows client and re-opening it (leaving the server
> running), the log then shows it read the nxagent configuration, and
> copy/paste works (and I can keep doing this, hence the log showing nxagent
> resuming twice). If I instead, logout (killing the server) and reconnect,
> it comes back as x2go (of course).
>

As I currently not able to replicate this: Can you downgrade again and
check? I guess your clipboard problems are not attached to the
nxagent/x2goagent mode but to initial/reconnected session state.

Uli

PS: we should better open two further issues instead of discussing those
three issues in one...

**Comment by andyleejordan** [comment]
> On Wed, May 8, 2019 at 12:41 AM Andrew Schwartzmeyer < ***@***.***> wrote: Session: Suspending session at 'Tue May 7 01:37:57 2019'. Info: Waiting the cleanup timeout to complete. Session: Session suspended at 'Tue May 7 01:37:58 2019'. Xlib: resource ID allocation space exhausted! Session: Resuming session at 'Tue May 7 21:58:41 2019'.
> I am wondering where that Xlib message is coming from. I have tried replicating that behaviour but failed so far... Will try further.
> When it suspends and resumes the session, it comes back up as nxagent, not x2go, and so reads the nx config files, not x2go's. ALSO, copy/paste for me is completely broken when it's the x2go agent, but if the session is suspended and resumed (and so it comes back as an nxagent), copy/paste works just fine. I can repro this by starting a new session, checking that the log has it using the x2go configuration, and that copy/paste does not work, and then just closing the Windows client and re-opening it (leaving the server running), the log then shows it read the nxagent configuration, and copy/paste works (and I can keep doing this, hence the log showing nxagent resuming twice). If I instead, logout (killing the server) and reconnect, it comes back as x2go (of course).
> As I currently not able to replicate this: Can you downgrade again and check? I guess your clipboard problems are not attached to the nxagent/x2goagent mode but to initial/reconnected session state. Uli PS: we should better open two further issues instead of discussing those three issues in one...

Yeah, the clipboard is not repro'ing consistently, so I think you're right.

> PS: we should better open two further issues instead of discussing those
three issues in one...

So true 😆

**Comment by uli42** [comment]
@andschwa: I have open issue #803 and PR #804 that should fix it. I have not found how I can add you to the issue so I write it here: if possible can you please apply that patch and check if it also fixes the clipboard error for you?

**Comment by andyleejordan** [comment]
Thank you @uli42! I will test that PR ASAP.

**Comment by uli42** [comment]
solved by #804
